Jellyfin.Server.Implementations/Events/EventManager.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Threading.Tasks;
+using MediaBrowser.Controller;
+using MediaBrowser.Controller.Events;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Jellyfin.Server.Implementations.Events
+{
+    /// <summary>
+    /// Handles the firing of events.
+    /// </summary>
+    public class EventManager : IEventManager
+    {
+        private readonly ILogger<EventManager> _logger;
+        private readonly IServerApplicationHost _appHost;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventManager"/> class.
+        /// </summary>
+        /// <param name="logger">The logger.</param>
+        /// <param name="appHost">The application host.</param>
+        public EventManager(ILogger<EventManager> logger, IServerApplicationHost appHost)
+        {
+            _logger = logger;
+            _appHost = appHost;
+        }
+
+        /// <inheritdoc />
+        public void Publish<T>(T eventArgs)
+            where T : EventArgs
+        {
+            Task.WaitAll(PublishInternal(eventArgs));
+        }
+
+        /// <inheritdoc />
+        public async Task PublishAsync<T>(T eventArgs)
+            where T : EventArgs
+        {
+            await PublishInternal(eventArgs).ConfigureAwait(false);
+        }
+
+        private async Task PublishInternal<T>(T eventArgs)
+            where T : EventArgs
+        {
+            using var scope = _appHost.ServiceProvider.CreateScope();
+            foreach (var service in scope.ServiceProvider.GetServices<IEventConsumer<T>>())
+            {
+                try
+                {
+                    await service.OnEvent(eventArgs).ConfigureAwait(false);
```

Comment:
In order to do that, there would have to be a separate scope resolved for each service, since certain services (EFCore) can't be used from multiple threads at once. Resolving separate scopes seemed more complicated than it was worth, so I chose to do this approach.